prov/verbs/src/fi_verbs.h

```diff
@@ -394,33 +443,177 @@ void fi_ibv_monitor_unsubscribe(struct ofi_mem_monitor *notifier,
+	struct fi_ibv_cq	*srq_cq;
+
+	/* XRC SRQ is not created until endpoint enable */
+	fastlock_t		prepost_lock;
+	ofi_fastlock_acquire_t	pp_fastlock_acquire;
+	ofi_fastlock_release_t	pp_fastlock_release;
+
+	struct slist		prepost_list;
+	uint32_t		max_recv_wr;
+	uint32_t		max_sge;
+	uint32_t		prepost_count;
 };
 
 int fi_ibv_srq_context(struct fid_domain *domain, struct fi_rx_attr *attr,
 		       struct fid_ep **rx_ep, void *context);
 
+/* Used to determine if XRC has been enabled and requested */
+#ifdef INCLUDE_VERBS_XRC
+static inline int fi_ibv_using_xrc(void)
+{
+	return fi_ibv_gl_data.msg.use_xrc;
+}
+
+static inline int fi_ibv_is_xrc_send_qp(enum ibv_qp_type qp_type)
+{
+	return qp_type == IBV_QPT_XRC_SEND;
+}
+#else /* INCLUDE_VERBS_XRC */
+static inline int fi_ibv_using_xrc(void)
+{
+	return 0;
+}
+static inline int fi_ibv_is_xrc_send_qp(enum ibv_qp_type qp_type)
+{
+	return 0;
+}
+#endif /* INCLUDE_VERBS_XRC */
+
+int fi_ibv_domain_xrc_init(struct fi_ibv_domain *domain);
+int fi_ibv_domain_xrc_cleanup(struct fi_ibv_domain *domain);
+
+enum fi_ibv_ini_qp_state {
+	FI_IBV_INI_QP_UNCONNECTED,
+	FI_IBV_INI_QP_CONNECTING,
+	FI_IBV_INI_QP_CONNECTED
+};
+
+#define FI_IBV_NO_INI_TGT_QPNUM 0
+#define FI_IBV_RECIP_CONN	1
+
+/*
+ * An XRC transport INI QP connection can be shared within a process to
+ * communicate with all the ranks on the same remote node. This structure is
+ * only accessed during connection setup and tear down and should be
+ * done while holding the domain:xrc:ini_mgmt_lock.
+ */
+struct fi_ibv_ini_shared_conn {
+	/* To share, EP must have same remote peer host addr and TX CQ */
+	struct sockaddr			*peer_addr;
+	struct fi_ibv_cq		*tx_cq;
+
+	/* The physical INI/TGT QPN connection. Virtual connections to the
+	 * same remote peer and TGT QPN will share this connection, with
+	 * the remote end opening the specified XRC TGT QPN for sharing. */
+	enum fi_ibv_ini_qp_state	state;
+	struct ibv_qp			*ini_qp;
+	uint32_t			tgt_qpn;
+
+	/* EP waiting on or using this INI/TGT physical connection will be in
+	 * one of these list and hold a reference to the shared connection. */
+	struct dlist_entry		pending_list;
+	struct dlist_entry		active_list;
+	ofi_atomic32_t			ref_cnt;
+};
+
+enum fi_ibv_xrc_ep_conn_state {
+	FI_IBV_XRC_UNCONNECTED,
+	FI_IBV_XRC_ORIG_CONNECTING,
+	FI_IBV_XRC_ORIG_CONNECTED,
+	FI_IBV_XRC_RECIP_CONNECTING,
+	FI_IBV_XRC_CONNECTED
+};
+
+/*
+ * The following XRC state is only required during XRC connection
+ * establishment and can be freed once bidirectional connectivity
+ * is established.
+ */
+struct fi_ibv_xrc_ep_conn_setup {
+	uint32_t			tag;
+
+	/* IB CM message stale/duplicate detection processing requires
+	 * that shared INI/TGT connections use unique QP numbers during
+	 * RDMA CM connection setup. To avoid conflicts with actual HCA
+	 * QP number space, we allocate minimal QP that are left in the
+	 * reset state and closed once the setup process completes. */
```

Comment:
I don't understand.  Is this a work-around for the rdma cm not supporting xrc?  Are we running unnecessary CM exchanges?  Is this using the CM protocol to exchange some other information via the CM private data?